PXC-512: Replication slave from PXC can crash if wsrep_forced_binlog_format=ROW

voodam opened the pull request:
https://jira.percona.com/browse/PXC-512

Problem:
Duplication of procedure on replication if `wsrep_forced_binlog_format=ROW`.

Cause:
`THD::set_current_stmt_binlog_format_row` and `THD::clear_current_stmt_binlog_format_row` methods inconsistent with `THD::is_current_stmt_binlog_format_row` (when `wsrep_forced_binlog_format` is enabled): `THD::is_current_stmt_binlog_format_row` returns value of `wsrep_forced_binlog_format`, but modifying methods update `THD::current_stmt_binlog_format` variable. So, when we temporary clear binlog format it continues to be forced set and this breaks replication logic.

Solution:
Update also `wsrep_forced_binlog_format` variable. For simplicity, global variable is updated for all sessions. Because `wsrep_forced_binlog_format` is deprecated this is considered enough.

Comment: probably this is dirty solution, but I tried for the first time assuming this can be enough.

Alternative approach, may be it will be better:
Create `THD::wsrep_forced_binlog_format_save` variable and modify it instead of global variable. Also it requires to pass `thd` into `WSREP_BINLOG_FORMAT`, because we need to check not only `wsrep_forced_binlog_format` but also `THD::wsrep_forced_binlog_format_save`.

it-percona-cla commented:
[![CLA assistant check](https://cla-assistant.percona.com:443/pull/badge/signed)](https://cla-assistant.percona.com:443/percona/percona-xtradb-cluster?pullRequest=1703) <br/>All committers have signed the CLA.

kamil-holubicki commented on sql/sql_class.h:
```diff
@@ -4421,12 +4421,28 @@ class THD :public MDL_context_owner,
   {
     DBUG_ENTER("clear_current_stmt_binlog_format_row");
+#ifdef WITH_WSREP
+    if (wsrep_forced_binlog_format != BINLOG_FORMAT_UNSPEC)
```
This will force setting `wsrep_forced_binlog_format` to `BINLOG_FORMAT_STMT`. Is it OK? Shouldn't we restore the original value?

voodam replied:
`clear_...` may be is not so good naming. Purpose of that method is set STMT format temporary. Original format is restored in `set_current_stmt_binlog_format_row`  method.

But it's right that there will be bug on restoring if `wsrep_forced_binlog_format != BINLOG_FORMAT_ROW`, I'll fix it soon.

kamil-holubicki reviewed (changes requested):
I think it will be better to go with the "alternative option".
`wsrep_forced_binlog_format` is the global variable, so it affects all sessions. What if we change it in a session that causes problems, but another parallel session would use it?
I would consider introducing a new session-bound member of THD (like `current_stmt_binlog_format`, e.g. THD::wsrep_forced_binlog_format_override). Then set it accordingly in `set_current_stmt_binlog_format_row()/clear_current_stmt_binlog_format_row()` to indicate that we internally forced the format regardless of `wsrep_forced_binlog_format`.
Of course, we will need to change the implementation of WSREP_BINLOG_FORMAT macro to be aware of `THD::wsrep_forced_binlog_format_override` (log.h and sql_class.h - BTW, would it be possible to get rid of one of them and have implementation only in one place?)

kamil-holubicki commented on sql/sql_class.h:
```diff
@@ -2638,6 +2640,24 @@ class THD :public MDL_context_owner,
   */
   enum_binlog_format current_stmt_binlog_format;
 
+  /**
```
This should go into #ifdef WITH_WSREP. Could we move it down to the existing section of WSREP-related members? (That will cause initialization order change in constructor)

Please consider changing the type of it to `enum_binlog_format`. 
Maybe changing the type of `wsrep_forced_binlog_format` is not a bad idea as well?

voodam replied:
1. Done #ifdef.
2. Changing type: all variables passed inside `WSREP_BINLOG_FORMAT` are ulong, so compiler throw warnings about mixed types. Is it worth to change the type in this case?

kamil-holubicki replied:
> 1. Done #ifdef.
> 2. Changing type: all variables passed inside `WSREP_BINLOG_FORMAT` are ulong, so compiler throw warnings about mixed types. Is it worth to change the type in this case?

OK, let's leave it as it is

kamil-holubicki commented on sql/sql_class.h:
```diff
@@ -2701,6 +2721,10 @@ class THD :public MDL_context_owner,
       skip_readonly_check= false;
   }
 
+  ulong wsrep_forced_binlog_format_override() const {
```
#ifdef WITH_WSREP

voodam replied:
Done

kamil-holubicki commented on sql/sql_class.h:
```diff
@@ -270,11 +270,13 @@ enum enum_binlog_format {
+       wsrep_forced_binlog_format : my_format))
 #else
-#define WSREP_BINLOG_FORMAT(my_format) my_format
+#define WSREP_BINLOG_FORMAT(thd, my_format) my_format
```
There is no WSREP_BINLOG_FORMAT if there is no WITH_WSREP. I know it was there before, but it is not necessary. Please remove 'else' part.

voodam replied:
Done

voodam replied:
Done

kamil-holubicki reviewed (changes requested):
Looks fine, just please consider refactoring as requested.

kamil-holubicki commented:
Please also rebase considering GCA process.

voodam commented:
> Please also rebase considering GCA process.

Probably changes are not required in 8.0, there are no `wsrep_forced_binlog_format` variable in that branch.

kamil-holubicki commented on sql/sql_class.h:
```diff
@@ -4421,12 +4444,20 @@ class THD :public MDL_context_owner,
   {
     DBUG_ENTER("clear_current_stmt_binlog_format_row");
+#ifdef WITH_WSREP
+    if (wsrep_forced_binlog_format != BINLOG_FORMAT_UNSPEC &&
```
One more question. Is this `if` needed at all? what if we just set
`m_wsrep_forced_binlog_format_override= BINLOG_FORMAT_STMT;` ?
It will work the same, right?

voodam replied:
The problem with `m_wsrep_forced_binlog_format_override` is that `wsrep_forced_binlog_format` can change and overrideing variable will become outdated. So I tried to set it to `BINLOG_FORMAT_UNSPEC` as soon as posible and don't change it without need.

Another problem, for example: `reset_current_stmt_binlog_format_row` is used in `THD::init`, so if `wsrep_forced_binlog_format` is not set and this isn't checked in if, override variable can be set in wrong value.

kamil-holubicki commented:
https://pxc.cd.percona.com/view/PXC%205.7/job/pxc-5.7-pipeline/4962/

voodam commented:
@kamil-holubicki, There is some problem in pipeline:
`[tempesta-tech/percona-xtradb-cluster/am-PXC-512-5.7/VERSION](https://raw.githubusercontent.com/tempesta-tech/percona-xtradb-cluster/am-PXC-512-5.7/VERSION)`
There is no such repository `tempesta-tech/percona-xtradb-cluster`, only `tempesta-tech/percona-server` exists.

kamil-holubicki commented:
> > Please also rebase considering GCA process.
> 
> Probably changes are not required in 8.0, there are no `wsrep_forced_binlog_format` variable in that branch.

Yes, but we still need to merge it to 8.0 (it will be empty merge)

voodam committed:
PXC-512: Replication slave from PXC can crash if wsrep_forced_binlog_format=ROW

https://jira.percona.com/browse/PXC-512

Problem:
Duplication of procedure on replication if wsrep_forced_binlog_format=ROW.

Cause:
THD::set_current_stmt_binlog_format_row and THD::clear_current_stmt_binlog_format_row methods are
inconsistent with THD::is_current_stmt_binlog_format_row when wsrep_forced_binlog_format is enabled:
THD::is_current_stmt_binlog_format_row returns value of wsrep_forced_binlog_format,
but modifying methods update THD::current_stmt_binlog_format variable. So, when we temporary
clear binlog format, it continues to be forced set and this breaks replication logic.

Solution:
Introduce THD::m_wsrep_forced_binlog_format_override variable that is set in modifying methods above,
and that has priority over global variable.

kamil-holubicki reviewed (approved):
LGTM. Please create 8.0 PR

voodam commented:
> LGTM. Please create 8.0 PR

https://github.com/percona/percona-xtradb-cluster/pull/1704

kamil-holubicki reviewed (approved):
lgtm
